src/types/UiaTextRangeBase.cpp

```diff
@@ -990,1534 +904,364 @@ const COORD UiaTextRangeBase::_getScreenFontSize() const
+// - endAnchor - the end anchor of interested data within the viewport. In text buffer coordinate space. Inclusive
+// - coords - vector to add the calculated coords to
 // Return Value:
-// - The row width
-const unsigned int UiaTextRangeBase::_getRowWidth(gsl::not_null<IUiaData*> pData)
+// - <none>
+void UiaTextRangeBase::_getBoundingRect(_In_ const COORD startAnchor, _In_ const COORD endAnchor, _Inout_ std::vector<double>& coords) const
 {
-    // make sure that we can't leak a 0
-    return std::max(static_cast<unsigned int>(_getScreenBufferCoords(pData).X), 1u);
-}
+    FAIL_FAST_IF(startAnchor.Y != endAnchor.Y);
 
-// Routine Description:
-// - calculates the column refered to by the endpoint.
-// Arguments:
-// - pData - the UiaData for the terminal we are operating on
-// - endpoint - the endpoint to translate
-// Return Value:
-// - the column value
-const Column UiaTextRangeBase::_endpointToColumn(gsl::not_null<IUiaData*> pData, const Endpoint endpoint)
-{
-    return endpoint % _getRowWidth(pData);
-}
+    const auto viewport = _pData->GetViewport();
+    const auto currentFontSize = _getScreenFontSize();
 
-// Routine Description:
-// - converts an Endpoint into its equivalent text buffer row.
-// Arguments:
-// - pData - the UiaData for the terminal we are operating on
-// - endpoint - the endpoint to convert
-// Return Value:
-// - the text buffer row value
-const TextBufferRow UiaTextRangeBase::_endpointToTextBufferRow(gsl::not_null<IUiaData*> pData,
-                                                               const Endpoint endpoint)
-{
-    return endpoint / _getRowWidth(pData);
-}
+    POINT topLeft{ 0 };
+    POINT bottomRight{ 0 };
 
-// Routine Description:
-// - counts the number of rows that are fully or partially part of the
-// range.
-// Arguments:
-// - pData - the UiaData for the terminal we are operating on
-// Return Value:
-// - The number of rows in the range.
-const unsigned int UiaTextRangeBase::_rowCountInRange(gsl::not_null<IUiaData*> pData) const
-{
-    if (_degenerate)
+    // startAnchor is converted to the viewport coordinate space
+#pragma warning(suppress : 26496) // analysis can't see this, TODO GH: 4015 to improve Viewport to be less bad because it'd go away if ConvertToOrigin returned instead of inout'd.
+    auto startCoord = startAnchor;
+    viewport.ConvertToOrigin(&startCoord);
+
+    // we want to clamp to a long (output type), not a short (input type)
+    // so we need to explicitly say <long,long>
+    topLeft.x = base::ClampMul<long, long>(startCoord.X, currentFontSize.X);
+    topLeft.y = base::ClampMul<long, long>(startCoord.Y, currentFontSize.Y);
+
+    if (IsDegenerate())
     {
-        return 0;
+        // An empty array is returned for a degenerate (empty) text range or for a text range
+        // reference: https://docs.microsoft.com/en-us/windows/win32/api/uiautomationclient/nf-uiautomationclient-iuiautomationtextrange-getboundingrectangles
+        return;
+    }
+    else

```

Comment:
you no longer need this and can remove the brace level. !Degenerate is a precondition of calling this function.

Reply:
Removed and added a comment at the top stating that assumption... And adding a FailFastIf, to enforce it I guess?

For future reference, is that generally the way to go? I always thought we should make a function as fail-proof as possible?

Then again, I guess adding FailFast's make sure we never misuse a function?